### PR TITLE
Use Http Chunking to stream logs to the client for tailing on logs.

### PR DESCRIPTION
**Pull request by brahmaroutu**
Fixes 3633

I plan to separate GetLog and GetLogStream so that we do not cause any regressions. GetLogStream and GetJobLogStream are two alternate paths currently executed if a 'follow=true' query parameter is added to the URL. It also honors other query parameters that can be passes to k8s. The 'follow=true' option only works against kubernetes, this option is ignored for other Jobs.

### CODE REVIEW

**Comment by spxtr** [comment]
See #4152 for a simplification that will affect this.

**Comment by BenTheElder** [comment]
Fixes https://github.com/kubernetes/test-infra/issues/3633 :tada:

Perhaps we can switch some of the links in EG Gubernator to default to this (eg the "raw log" link?).
It would be cool if we could get streaming results by default, and further incentive to move off of jenkins :smile:

File: prow/cmd/deck/main_test.go [hunk 1]
```diff
@@ -82,6 +103,11 @@ func TestHandleLog(t *testing.T) {
 			code: http.StatusOK,
 		},
 		{
+			name: "pod that does exist",
+			path: "?pod=pn;follow=true",
```
Comment (rmmh) [inline prow/cmd/deck/main_test.go]: use & to separate parameters

Comment (brahmaroutu) [inline prow/cmd/deck/main_test.go]: Thanks for the review, replaces as suggested.

File: prow/cmd/deck/main.go [hunk 2]
```diff
@@ -116,6 +121,36 @@ func handleData(ja *JobAgent) http.HandlerFunc {
...
+	reader := bufio.NewReader(log)
+	for {
+		line, err := reader.ReadBytes('\n')
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+		}
+		cw.Write(line)
+		flusher.Flush()
```
Comment (rmmh) [inline prow/cmd/deck/main.go]: wrap this in an `if ok {`, since the http.Flusher coercion can fail?

File: prow/cmd/deck/main.go [hunk 3]
```diff
@@ -116,6 +121,36 @@ func handleData(ja *JobAgent) http.HandlerFunc {
 
 type logClient interface {
 	GetJobLog(job, id string) ([]byte, error)
+	GetJobLogStream(job, id string, options map[string]string) (io.ReadCloser, error)
```
Comment (rmmh) [inline prow/cmd/deck/main.go]: we should

1) Document that options are passed straight through to the Kubernetes [read log](https://kubernetes.io/docs/api-reference/v1.6/#read-log) API.
2) Decide if that makes sense. Should we instead just pass follow=true? Does passing these parameters through allow better frontends?

File: prow/cmd/deck/main.go [hunk 4]
```diff
@@ -125,6 +160,15 @@ func handleLog(lc logClient) http.HandlerFunc {
 		w.Header().Set("Access-Control-Allow-Origin", "*")
 		job := r.URL.Query().Get("job")
 		id := r.URL.Query().Get("id")
+		stream := r.URL.Query().Get("follow")
+		var logStreamRequested bool
+		if ok, _ := strconv.ParseBool(stream); ok {
+			// get http chunked responses to the client
+			w.Header().Set("Connection", "Keep-Alive")
+			w.Header().Set("Transfer-Encoding", "chunked")
+			w.Header().Set("X-Content-Type-Options", "nosniff")
```
Comment (rmmh) [inline prow/cmd/deck/main.go]: Why are Connection and X-Content-Type-Options needed here?

File: prow/cmd/deck/main_test.go [hunk 5]
```diff
@@ -87,18 +97,42 @@ func TestHandleLog(t *testing.T) {
...
 		} else if rr.Code == http.StatusOK {
-			resp := rr.Result()
-			defer resp.Body.Close()
-			if body, err := ioutil.ReadAll(resp.Body); err != nil {
-				t.Errorf("Error reading response body: %v", err)
-			} else if string(body) != "hello" {
-				t.Errorf("Unexpected body: got %s.", string(body))
+			if follow {
+				//wait a little to get the chunks
+				time.Sleep(2 * time.Millisecond)
```
Comment (rmmh) [inline prow/cmd/deck/main_test.go]: Why is this necessary? Shouldn't reader.readBytes('\n') block while waiting anyways?

File: prow/cmd/deck/main.go (hunk 2 above)
Comment (brahmaroutu) [inline prow/cmd/deck/main.go]: Not sure what you are saying, Flush() on flusher does not return error.

File: prow/cmd/deck/main.go (hunk 3 above)
Comment (brahmaroutu) [inline prow/cmd/deck/main.go]: I agree that it is a design choice, I will wait for opinions. I like to have the ability to have all options enabled.

File: prow/cmd/deck/main.go (hunk 4 above)
Comment (brahmaroutu) [inline prow/cmd/deck/main.go]: Fixed.

File: prow/cmd/deck/main_test.go (hunk 5 above)
Comment (brahmaroutu) [inline prow/cmd/deck/main_test.go]: Test consistently gets no data if I don't wait. Simplest way to wait is sleep.

File: prow/cmd/deck/main.go [hunk 6]
```diff
@@ -116,6 +121,39 @@ func handleData(ja *JobAgent) http.HandlerFunc {
...
 type logClient interface {
 	GetJobLog(job, id string) ([]byte, error)
+	// Add ability to stream logs with options enabled. This call is used to follow logs
+	// using kubernetes client API. All other options on the Kubernetes log api can
+	// also be enabled.
+	GetJobLogStream(job, id string, options map[string]string) (io.ReadCloser, error)
+}
+
+func httpChunking(log io.ReadCloser, w http.ResponseWriter) {
+	flusher, ok := w.(http.Flusher)
```
Comment (rmmh) [inline prow/cmd/deck/main.go]: flusher will be a nil interface if this coercion fails, meaning that flusher.Flush() will panic

Comment (brahmaroutu) [reply prow/cmd/deck/main.go]: thanks, I am now checking for nil.

File: prow/cmd/deck/jobs.go [hunk 7]
```diff
@@ -119,6 +121,24 @@ func (ja *JobAgent) GetJobLog(job, id string) ([]byte, error) {
 	return ja.jc.GetLog(job, num)
 }
 
+func (ja *JobAgent) GetJobLogStream(job, id string, options map[string]string) (io.ReadCloser, error) {
```
Comment (rmmh) [inline prow/cmd/deck/jobs.go]: please document that options is query parameters documented here: https://kubernetes.io/docs/api-reference/v1.7/#read-log

**Comment by kargakis** [comment]
What's really needed to support this for Jenkins, too?

cc @stevekuznetsov

**Comment by stevekuznetsov** [comment]
We'd have to check that the equivalent API endpoint exists in Jenkins.
